src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.cs

```diff
@@ -0,0 +1,718 @@
+
+            // The rest of the argument validation is done by CopyTo
+
+            new Span<T>(source, startIndex, length).CopyTo(new Span<T>((void*)destination, length));
+        }
+
+        public static void Copy(IntPtr source, int[] destination, int startIndex, int length)
+        {
+            CopyToManaged(source, destination, startIndex, length);
+        }
+
+        public static void Copy(IntPtr source, char[] destination, int startIndex, int length)
+        {
+            CopyToManaged(source, destination, startIndex, length);
+        }
+
+        public static void Copy(IntPtr source, short[] destination, int startIndex, int length)
+        {
+            CopyToManaged(source, destination, startIndex, length);
+        }
+
+        public static void Copy(IntPtr source, long[] destination, int startIndex, int length)
+        {
+            CopyToManaged(source, destination, startIndex, length);
+        }
+
+        public static void Copy(IntPtr source, float[] destination, int startIndex, int length)
+        {
+            CopyToManaged(source, destination, startIndex, length);
+        }
+
+        public static void Copy(IntPtr source, double[] destination, int startIndex, int length)
+        {
+            CopyToManaged(source, destination, startIndex, length);
+        }
+
+        public static void Copy(IntPtr source, byte[] destination, int startIndex, int length)
+        {
+            CopyToManaged(source, destination, startIndex, length);
+        }
+
+        public static void Copy(IntPtr source, IntPtr[] destination, int startIndex, int length)
+        {
+            CopyToManaged(source, destination, startIndex, length);
+        }
+
+        private static unsafe void CopyToManaged<T>(IntPtr source, T[] destination, int startIndex, int length)
+        {
+            if (source == IntPtr.Zero)
+                throw new ArgumentNullException(nameof(source));
+            if (destination == null)
+                throw new ArgumentNullException(nameof(destination));
+            if (startIndex < 0)
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_StartIndex);
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length), SR.ArgumentOutOfRange_NeedNonNegNum);
+
+            // The rest of the argument validation is done by CopyTo
+
+            new Span<T>((void*)source, length).CopyTo(new Span<T>(destination, startIndex, length));
+        }
+
+        public static unsafe byte ReadByte(IntPtr ptr, int ofs)
+        {
+            try
+            {
+                byte* addr = (byte*)ptr + ofs;
+                return *addr;
+            }
+            catch (NullReferenceException)
+            {
+                // this method is documented to throw AccessViolationException on any AV
+                throw new AccessViolationException();
+            }
```

Comment:
Rather than throwing/catching and having a try/catch that to my knowledge impedes some JIT optimizations, would it be better to do:
```C#
byte* addr = (byte*)ptr + ofs;
if (addr == null)
{
    throw new AccessViolationException();
}
return *addr;
```
? Or is there something else here that could cause a NullReferenceException?

Reply:
Small offsets will turn into NullReferenceException too, so this would need to be `if (addr < FirstPageSize)` if we wanted to maintain the behavior.

These APIs have a lot of legacy baggage with them. I would leave it on the `try / catch` plan.

Reply:
Ok.